### PR TITLE
feat: auto generation EHF and spork+EHF activation for MN_RR

### PR DESCRIPTION
**Pull request by knst**
Implementation EHF mechanism, part 4. Previous changes are: 
 - https://github.com/dashpay/dash/pull/4577
 - https://github.com/dashpay/dash/pull/5505
 - https://github.com/dashpay/dash/pull/5469

## Issue being fixed or feature implemented
Currently MN_RR is activated automatically by soft-fork activation after v20 is activated.
It is not flexible enough, because platform may not be released by that time yet or in opposite it can be too long to wait.
Also, any signal of EHF requires manual actions from MN owners to sign EHF signal - it is automated here.

## What was done?
New spork `SPORK_24_MN_RR_READY`; new EHF manager that sign EHF signals semi-automatically without manual actions; and send transaction with EHF signal when signal is signed to network. 
Updated rpc `getblockchaininfo` to return information about of EHF activated forks.
Fixed function `IsTxSafeForMining` in chainlock's handler to skip transactions without inputs (empty `vin`).

## How Has This Been Tested?
Run unit/functional tests. Some tests have been updated due to new way of MN_RR activation: `feature_asset_locks.py`, `feature_mnehf.py`, `feature_llmq_evo.py` and unit test `block_reward_reallocation_tests`.

## Breaking Changes
New way of MN_RR activation.

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
pls see 2ccf9d1825

**Comment by knst** [comment]
@UdjinM6 
A.
https://github.com/dashpay/dash/commit/2ccf9d1825b41d663701ce92c0a5b01a44eb6562#diff-4a04bc0b355c780033960e8c261ee9b6d3c452897e1dcd88a15d272512266c76R1133

> self.activate_mn_rr(expected_activation_height=node.getblockcount() + 12 * 3)

but it can take a time until recovered sigs are ready and tx with EHF signal is submitted more than one loop iteration:

```
       while mn_rr_status == 0:
            time.sleep(1)
            mn_rr_status = get_bip9_details(self.nodes[0], 'mn_rr')['ehf']
            self.nodes[0].generate(1)
```

So, if sig for EHF signal is delayed more than 1 second and 1 block -> it may be more that 12 * 3 blocks required to activate mn_rr, isn't it?


----
B.
consensus.vDeployments[Consensus::DEPLOYMENT_MN_RR].nWindowSize = 4032;
When v20 would be released it will take time until v20 would be activated (several weeks, 4-6?). After that after spork activation it would be required to wait 4-6 weeks more. Is it expected to wait 4-6 weeks more?

**Comment by UdjinM6** [comment]
> @UdjinM6 A. [2ccf9d1#diff-4a04bc0b355c780033960e8c261ee9b6d3c452897e1dcd88a15d272512266c76R1133](https://github.com/dashpay/dash/commit/2ccf9d1825b41d663701ce92c0a5b01a44eb6562#diff-4a04bc0b355c780033960e8c261ee9b6d3c452897e1dcd88a15d272512266c76R1133)
> 
> > self.activate_mn_rr(expected_activation_height=node.getblockcount() + 12 * 3)
> 
> but it can take a time until recovered sigs are ready and tx with EHF signal is submitted more than one loop iteration:
> 
> ```
>        while mn_rr_status == 0:
>             time.sleep(1)
>             mn_rr_status = get_bip9_details(self.nodes[0], 'mn_rr')['ehf']
>             self.nodes[0].generate(1)
> ```
> 
> So, if sig for EHF signal is delayed more than 1 second and 1 block -> it may be more that 12 * 3 blocks required to activate mn_rr, isn't it?

bip9 works in periods. Also, in this specific case we start activating at the beginning of the period (block 2796) and we have threshold which is lower than that so missing a couple of blocks is not a problem. But I agree that calculating it like that is probably misleading. We could simply go with a static number instead I guess i.e. `expected_activation_height=2832`.

> B. consensus.vDeployments[Consensus::DEPLOYMENT_MN_RR].nWindowSize = 4032; When v20 would be released it will take time until v20 would be activated (several weeks, 4-6?). After that after spork activation it would be required to wait 4-6 weeks more. Is it expected to wait 4-6 weeks more?

Good point 👍  Pls ignore this part of my suggestions then :)

**Review by UdjinM6 — APPROVED** [review]
lgtm, utACK

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK

File: src/evo/mnhftx.h [hunk 1]
```diff
@@ -63,6 +63,17 @@ class MNHFTxPayload
     uint8_t nVersion{CURRENT_VERSION};
     MNHFTx signal;
 
+public:
+    /**
+     * helper function to calculare Request ID used for signing
```
Comment (PastaPastaPasta) [inline src/evo/mnhftx.h:68]:
```suggestion
     * helper function to calculate Request ID used for signing
```

**Review by UdjinM6 — APPROVED** [review]
re-utACK

File: src/llmq/ehf_signals.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+{
+private:
+    CChainState& chainstate;
+    CConnman& connman;
+    CSigningManager& sigman;
+    CSigSharesManager& shareman;
+    CSporkManager& sporkman;
+    CQuorumManager& qman;
+    CTxMemPool& mempool;
+    CMNHFManager& mnhfManager;
```
Comment (PastaPastaPasta) [inline src/llmq/ehf_signals.h:35]: These can be const (and change constructor)
```suggestion
    const CSporkManager& sporkman;
    const CQuorumManager& qman;
    CTxMemPool& mempool;
    const CMNHFManager& mnhfManager;
```

**Review by UdjinM6 — DISMISSED** [review]
LGTM, utACK

**Review by PastaPastaPasta — APPROVED** [review]
light-ACK for squash merge

**Comment by knst** [comment]
> @[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/b6b165fc43e6bafed3e48cb0c56fd40ab3dd253d..7e96c96a60c06ab0692a4a419cfd92015923e75f) the mnehf-manager branch from[ b6b165f ](https://github.com/dashpay/dash/commit/b6b165fc43e6bafed3e48cb0c56fd40ab3dd253d)to 7e96c96

due to
>This pull request has conflicts, please rebase.

File: test/functional/feature_asset_locks.py [hunk 3]
```diff
@@ -515,7 +519,7 @@ def run_test(self):
         all_mn_rewards = platform_reward + owner_reward + operator_reward
         assert_equal(all_mn_rewards, bt['coinbasevalue'] * 3 // 4)  # 75/25 mn/miner reward split
         assert_equal(platform_reward, all_mn_rewards * 375 // 1000)  # 0.375 platform share
-        assert_equal(platform_reward, 25553999)
+        assert_equal(platform_reward, 29636590)
```
Comment (UdjinM6) [inline test/functional/feature_asset_locks.py:522]: somehow this test is not 100% deterministic....
run 1
```
TestFramework (INFO): Wait for mn_rr activation
TestFramework (INFO): height: 2855 credit: 897274252983
```
run 2
```
TestFramework (INFO): Wait for mn_rr activation
TestFramework (INFO): height: 2831 credit: 895835213008
```
and it affects `platform_reward` too so maybe we could rely on formulas only here?
```suggestion
```

Comment (UdjinM6) [reply test/functional/feature_asset_locks.py:522]: ok, so it happened only once for me and I can no longer reproduce it... nvm then

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
re-utACK

**Pull request merged**